File: go/triangle/triangle.go
// Package triangle is a utility to determine if a triangle is equilateral, isosceles, or scalene.
package triangle

// Kind represents the type of triangle or not a triangle.
type Kind int

const (
	NaT Kind = iota // not a triangle
	Equ             // equilateral
	Iso             // isosceles
	Sca             // scalene
)

// KindFromSides returns the kind of triangle formed by the given three sides.
func KindFromSides(a, b, c float64) Kind {
	switch {
	case !valid(a, b, c):
		return NaT

	}
}

func valid(a, b, c float64) bool {
	return a+b >= c && b+c >= a && a+c >= b
}

File: fsharp/triangle/Triangle.fs
module Triangle

let private triangleWithDistinctSides sides triangle =
    match triangle with
    | [ a; b; c ] -> a > 0. && b > 0. && c > 0. && a + b >= c && b + c >= a && a + c >= b
    | _ -> false
    && List.exists (fun n -> triangle |> List.distinct |> List.length = n) sides

let equilateral = triangleWithDistinctSides (List.singleton 1)

let isosceles = triangleWithDistinctSides ([ 1; 2 ])

let scalene = triangleWithDistinctSides (List.singleton 3)
